File: LiTCoder/Mod2/Mod2_Lab_2_ii.py
MOD = 1000000007

class LegoBlocks:
    def __init__(self, m, n):
        self.M = m
        self.N = n
        self.table = [-1] * (n + 1)
        self.table2 = [-1] * (n + 1)
        self.fill_table2(n)
        
        for i in range(n + 1):
            res = 1
            for j in range(m):
                res = (res * self.table2[i]) % MOD
            self.table2[i] = res

    def fill_table2(self, n):
        if n < 0:
            return 0
        if self.table2[n] == -1:
            if n == 0:
                self.table2[n] = 1
            else:
                self.table2[n] = (self.fill_table2(n - 1) + self.fill_table2(n - 2) +
                                  self.fill_table2(n - 3) + self.fill_table2(n - 4)) % MOD
        return self.table2[n]

    def evaluate(self):
        result = self.helper(self.N)
        return result

    def helper(self, n):
        if self.table[n] == -1:
            if n == 1:
                self.table[n] = 1
            else:
                self.table[n] = self.table2[n]
                for i in range(1, n):
                    self.table[n] = (self.table[n] - self.helper(n - i) * self.table2[i]) % MOD
                if self.table[n] < 0:
                    self.table[n] += MOD
        return self.table[n]

if __name__ == "__main__":
    m, n = map(int, input().split())
    lego_blocks = LegoBlocks(m, n)
    print(lego_blocks.evaluate())

File: LiTCoder/Mod2/Mod2_Lab2_ii.cpp
#include <iostream>
#include <vector>

#define MOD 1000000007

class LegoBlocks {
public:
    LegoBlocks(int m, int n) : M(m), N(n), table(n + 1, -1), table2(n + 1, -1) {
        fill_table2(n);

        for (int i = 0; i <= n; ++i) {
            int res = 1;
            for (int j = 0; j < m; ++j) {
                res = (res * table2[i]) % MOD;
            }
            table2[i] = res;
        }
    }

    int evaluate() {
        return helper(N);
    }

private:
    int M, N;
    std::vector<int> table;
    std::vector<int> table2;

    int fill_table2(int n) {
        if (n < 0) {
            return 0;
        }
        if (table2[n] == -1) {
            if (n == 0) {
                table2[n] = 1;
            } else {
                table2[n] = (fill_table2(n - 1) + fill_table2(n - 2) + fill_table2(n - 3) + fill_table2(n - 4)) % MOD;
            }
        }
        return table2[n];
    }

    int helper(int n) {
        if (table[n] == -1) {
            if (n == 1) {
                table[n] = 1;
            } else {
                table[n] = table2[n];
                for (int i = 1; i < n; ++i) {
                    table[n] = (table[n] - helper(n - i) * table2[i]) % MOD;
                }
                if (table[n] < 0) {
                    table[n] += MOD;
                }
            }
        }
        return table[n];
    }
};

int main() {
    int m, n;
    std::cin >> m >> n;

    LegoBlocks lego_blocks(m, n);
    std::cout << lego_blocks.evaluate() << std::endl;

    return 0;
}